File: MGI_analytique.py
import numpy as np

# Variables globales à définir ailleurs
L1 = None 
L2 = None 
Rb = None 
Re = None 

def MGI_analytique(eff):
    global L1, L2, Rb, Re

    # Matrice de rotation + translation de l'effecteur
    theta = eff[2]
    RotEff = np.array([
        [np.cos(theta), -np.sin(theta)],
        [np.sin(theta),  np.cos(theta)]
    ])
    Transl = np.array([[eff[0]], [eff[1]]])
    THEff = np.block([
        [RotEff, Transl],
        [np.zeros((1, 2)), [1]]
    ])

    ang1 = [0, 2 * np.pi / 3, 4 * np.pi / 3]
    ang2 = [-np.pi / 2, np.pi / 6, 5 * np.pi / 6]

    q = []

    for i in range(3):
        # Transformation homogène du repère R_i
        Rot = np.array([
            [np.cos(ang1[i]), -np.sin(ang1[i])],
            [np.sin(ang1[i]),  np.cos(ang1[i])]
        ])
        TH = np.block([
            [Rot, np.array([[Rb * np.cos(ang2[i])], [Rb * np.sin(ang2[i])]])],
            [np.zeros((1, 2)), [1]]
        ])

        # Point Ei dans le repère effecteur
        PEi_E = np.array([
            [Re * np.cos(ang2[i])],
            [Re * np.sin(ang2[i])],
            [1]
        ])

        # Position de Ei dans le repère monde
        PEi_0 = THEff @ PEi_E

        # Position de Ei dans le repère R_i
        PEi_i = np.linalg.inv(TH) @ PEi_0
        x, y = PEi_i[0, 0], PEi_i[1, 0]

        # Cinématique inverse 2R
        aux = (x**2 + y**2 - L1**2 - L2**2) / (2 * L1 * L2)
        if abs(aux) < 1:
            beta = np.arccos(aux)  # solution coude en haut
        else:
            beta = 0
            print("⚠️ Problème d'atteignabilité")

        alpha = np.arctan2(y, x) - np.arctan2(L2 * np.sin(beta), L1 + L2 * np.cos(beta))
        q.append([alpha, beta])

    return np.array(q)

File: MGI_analytique.m
function q=MGI_analytique(eff)

global L1 L2 Rb Re

%Mat. rotation et TH de l'effecteur
RotEff=[cos(eff(3)) -sin(eff(3)); sin(eff(3)) cos(eff(3))];
Transl=[eff(1); eff(2)];
THEff=[RotEff Transl; 0 0 1];

%angle R_i par rapport R_0
ang1=[0 2*pi/3 4*pi/3];

%angle des positions O_i et E_i 
ang2=[-pi/2 pi/6 5*pi/6];

q=[];

for i=1:3
    %Mat Rot et TH de R_i par rapport à R_0
    Rot=[cos(ang1(i)) -sin(ang1(i)); sin(ang1(i)) cos(ang1(i))];
    TH=[Rot [Rb*cos(ang2(i)); Rb*sin(ang2(i))]; [0 0 1]];
    
    %Position des points E_i dans R_E
    PEi_E=[Re*cos(ang2(i)); Re*sin(ang2(i)); 1];
    
    %Position des trois points E_i de l'effecteur dans R_0
    PEi_0=THEff*PEi_E;

    % Position des points effecteur E_i dans les repères dans robots R_i
    PEi_i=inv(TH)*PEi_0;

    %MGI 2R plan
    x=PEi_i(1);
    y=PEi_i(2);

    aux=(x^2+y^2-L1^2-L2^2)/(2*L1*L2);
    if abs(aux) < 1 
        beta=acos(aux); % changer le signe pour la solution coude en haut
    else
        beta=0; disp('problème d atteignabilité');
    end
    alpha=atan2(y,x)-atan2(L2*sin(beta),L1+L2*cos(beta));
    q=[q;[alpha; beta]];
    
end



end
